### PR TITLE
[BEAM-328] uses SerializableCoder for `T extends Serializable`

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @kennknowles

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 1]
```diff
@@ -720,6 +722,9 @@ private CoderFactory getDefaultCoderFactory(Class<?> clazz) throws CannotProvide
       return getDefaultCoder(clazz);
     } else if (type instanceof ParameterizedType) {
       return getDefaultCoder((ParameterizedType) type, typeCoderBindings);
+    } else if (type instanceof TypeVariable
+        && TypeToken.of(type).isSubtypeOf(Serializable.class)) {
+      return SerializableCoder.PROVIDER.getCoder(TypeDescriptor.of(type));
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: This is a bit overspecialized. What I think makes sense is get the `TypeDescriptor` to the `fallbackCoderProvider` on [line 336](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java#L336). I think you just need to adjust the overload that is used on [line 722](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java#L722) - it is just a historical accident that it only accepts a `Class`. (you might need to adjust names and `public`/`private`, etc)

**Comment by kennknowles** [comment]
Thanks for digging into this!

**Comment by manuzhang** [comment]
@kennknowles 
I found it simpler to use `TypeDescriptor.of(type).getRawType()` although we lose the class info in the error message for non `Serializable`

**Comment by manuzhang** [comment]
@kennknowles 
Oh, that breaks UTs.  Shall we throw exceptions with  `ResonCode.TYPE_ERASURE` for non `Serializable` types ?

**Comment by kennknowles** [comment]
For non-`Serializable` types, the fallback coder provider will throw `CannotProvideCoderException` and the coder registry will move on to the next attempt to provide a coder.

**Comment by dhalperi** [comment]
Hey Manu, the Jenkins test failed with a flake. Would you mind rebasing on master and pushing?

**Comment by manuzhang** [comment]
@dhalperi rebased and passed this time

**Comment by kennknowles** [comment]
This looks great. It could help many users. Merging.

**Pull request merged**